weldx/utility.py

```diff
@@ -562,6 +562,148 @@ def xr_interp_like(
+    ...         "d3": ["x", "y", "z"],
+    ...     }
+    ... )
+    >>> ref = dict(
+    ...     d1={"optional": True, "values": np.array([-1, 0, 2], dtype=int)},
+    ...     d2={
+    ...         "values": pd.DatetimeIndex(["2020-05-01", "2020-05-03"]),
+    ...         "dtype": ["datetime64[ns]", "timedelta64[ns]"],
+    ...     },
+    ...     d3={"values": ["x", "y", "z"], "dtype": "<U1"},
+    ... )
+    >>> wx.utility.xr_check_coords(dax, ref)
+    True
+
+    """
+    # check if the keys in ref are also in dax
+    xr_valid_key(dax, ref)
+
+    for key in ref:
+        # check if the optional key is set to true
+        if "optional" in ref[key]:
+            if ref[key]["optional"] and not hasattr(dax, key):
+                # skip this key - it is not in dax
+                continue
+
+        # only if the key "values" is given do the validation
+        if "values" in ref[key]:
+            if not (getattr(dax, key).values == ref[key]["values"]).all():
+                raise Exception(f"Value mismatch in DataArray and ref['{key}']")
+
+        # only if the key "dtype" is given do the validation
+        if "dtype" in ref[key]:
+            if isinstance(ref[key]["dtype"], list):
+                if [
+                    x
+                    for x in [x for x in ref[key]["dtype"] if isinstance(x, str)]
+                    if "timedelta64" in x or "datetime64" in x
+                ]:
+                    if np.issubdtype(
+                        getattr(dax, key).dtype, np.dtype("timedelta64")
+                    ) or np.issubdtype(getattr(dax, key).dtype, np.dtype("datetime64")):
+                        continue
+
+                if getattr(dax, key).dtype not in [
+                    np.dtype(x) for x in ref[key]["dtype"]
+                ]:
+                    if not (
+                        str in [np.dtype(x) for x in ref[key]["dtype"]]
+                        and np.issubdtype(getattr(dax, key).dtype, np.str_)
+                    ):
+                        raise Exception(
+                            f"Mismatch in the dtype of the DataArray and ref['{key}']"
+                        )
+            elif getattr(dax, key).dtype != np.dtype(ref[key]["dtype"]):
+                if isinstance(ref[key]["dtype"], str):
+                    if (
+                        "timedelta64" in ref[key]["dtype"]
+                        or "datetime64" in ref[key]["dtype"]
+                    ):
+                        if np.issubdtype(
+                            getattr(dax, key).dtype, np.dtype(ref[key]["dtype"])
+                        ):
+                            continue
+
+                if not (
+                    np.issubdtype(getattr(dax, key).dtype, np.dtype(ref[key]["dtype"]))
+                    and np.dtype(ref[key]["dtype"]) == str
+                ):
+                    raise Exception(
+                        f"Mismatch in the dtype of the DataArray and ref['{key}']"
+                    )
```

Comment:
This is too complex and too much code duplication.

Please write a single function that tests a coordinate dtype against a single input like:
~~~python
def _check_dtype(var_dtype,ref_dtype) -> bool:
    """Checks if dtype mtaches a reference dtype.

    Returns True or False"""

# use it like this
_check_dtype(getattr(dax, key).dtype,"datetime64")
_check_dtype(getattr(dax, key).dtype,str)
# etc.

# if list of allowed dtypes is passed:
any(_check_dtype(getattr(dax, key).dtype,var_dtype) for var_dtype in list_of_dtypes)
~~~